borg/_hashindex.c

```diff
@@ -220,6 +230,19 @@ int shrink_size(int current){
     return hash_sizes[i];
 }
 
+int
+count_empty(HashIndex *index)
+{   /* count empty (never used) buckets. this does NOT include deleted buckets (tombstones).
+     * TODO: if we ever change HashHeader, save the count there so we do not need this function.
```

Comment:
mmaping only makes sense if you advise the OS to actually load the data, otherwise there will be a large warmup period where you'll get a page fault for many accesses and have to block for essentially random-looking IO (since it's a hashtable...).

So, besides some minor advantages w.r.t. memory usage, mmaping can mainly make things faster that *don't* use the data structure much, say, a "borg list" that just retrieves a single item from the repository index.